File: src/Functional/Elixir/prototype.exs
defmodule Prototype do
  def clone_profile(profile) do
    %{profile | features: Enum.map(profile.features, & &1)}
  end

  def describe(profile) do
    "#{profile.name}: #{Enum.join(profile.features, ",")}"
  end
end

original = %{name: "orders", features: ["metrics"]}
base_clone = Prototype.clone_profile(original)
canary = %{base_clone | name: "orders-canary", features: base_clone.features ++ ["tracing"]}

IO.puts("original=#{Prototype.describe(original)}")
IO.puts("clone=#{Prototype.describe(canary)}")

File: src/Functional/Erlang/prototype.erl
-module(prototype).
-export([main/0]).

clone_profile(Profile) ->
    Profile#{features := lists:map(fun(X) -> X end, maps:get(features, Profile))}.

describe(Profile) ->
    Name = maps:get(name, Profile),
    Features = string:join(maps:get(features, Profile), ","),
    Name ++ ": " ++ Features.

main() ->
    Original = #{name => "orders", features => ["metrics"]},
    BaseClone = clone_profile(Original),
    Canary = BaseClone#{name := "orders-canary", features := maps:get(features, BaseClone) ++ ["tracing"]},
    io:format("original=~s~n", [describe(Original)]),
    io:format("clone=~s~n", [describe(Canary)]).
